src/main/resources/xsl/openagrar-solr.xsl

```diff
@@ -49,30 +49,62 @@
-        <field name="mods.mainAuthor">
-          <xsl:for-each select="mods:displayForm | mods:namePart | text()">
-            <xsl:value-of select="concat(' ',mcrxsl:normalizeUnicode(.))" />
-          </xsl:for-each>
-        </field>
-        <field name="mods.mainAuthor.affiliation">
-          <xsl:value-of select="mods:affiliation" />
-        </field>
-      </xsl:if>
-    </xsl:for-each>
-    <xsl:for-each
-      select="mods:name[mods:role/mods:roleTerm[@authority='marcrelator' and (@type='text' and text()='author') or (@type='code' and text()='aut')]]">
-      <xsl:if test="position()= last()">
-        <field name="mods.lastAuthor">
-          <xsl:for-each select="mods:displayForm | mods:namePart | text()">
-            <xsl:value-of select="concat(' ',mcrxsl:normalizeUnicode(.))" />
+      <xsl:choose>
+        <xsl:when test="not(../mods:name/mods:role/mods:roleTerm[contains(@valueURI, '#')])">
+          <xsl:variable name="autrole">
+            <xsl:choose>
+              <xsl:when test="position()=1">
+                <xsl:text>mainAuthor</xsl:text>
+              </xsl:when>	
+              <xsl:when test="position()=last()">
+                <xsl:text>lastAuthor</xsl:text>
+              </xsl:when>
+              <xsl:otherwise>
+               <xsl:text>noRole</xsl:text>
+              </xsl:otherwise>
+            </xsl:choose>
+          </xsl:variable>						
+          <xsl:if test="$autrole != 'noRole'">
+            <field name="{concat('mods.', $autrole)}">
+              <xsl:for-each select="mods:displayForm | mods:namePart | text()">
+                <xsl:value-of select="concat(' ',mcrxsl:normalizeUnicode(.))" />
+              </xsl:for-each>
+            </field>
+            <field name="{concat('mods.', $autrole, '.affiliation')}">
+              <xsl:value-of select="mods:affiliation" />
+            </field>
+          </xsl:if>
+        </xsl:when>
+        <xsl:otherwise>
+          <xsl:for-each select="mods:role/mods:roleTerm[contains(@valueURI, '#')]">
+            <xsl:variable name="autrole">
+              <xsl:choose>
+                <xsl:when test="substring-after(@valueURI, '#') ='main_author'">
+                  <xsl:text>mainAuthor</xsl:text>
+                </xsl:when>
+                <xsl:when test="substring-after(@valueURI, '#') ='co_author'">
+                  <xsl:text>coAuthor</xsl:text>
+                </xsl:when>
+                <xsl:when test="substring-after(@valueURI, '#') ='corresponding_author'">
+                  <xsl:text>correspondingAuthor</xsl:text>
+                </xsl:when>
+                <xsl:when test="substring-after(@valueURI, '#') ='last_author'">
+                  <xsl:text>lastAuthor</xsl:text>
+                </xsl:when>
+              </xsl:choose>
+            </xsl:variable>
+            <field name="{concat('mods.', $autrole)}">
+              <xsl:for-each select="../../mods:displayForm | ../../mods:namePart | ../../text()">
```

Comment:
In dem Solrfeld werden die Namen doppelt geschrieben. Aus 

```
<mods:name type="personal" xlink:type="simple">
<mods:displayForm>Herbst, Werner</mods:displayForm>
<mods:role>
<mods:roleTerm authority="marcrelator" type="code">aut</mods:roleTerm>
<mods:roleTerm authorityURI="https://www.mycore.de/classifications/author_roles" valueURI="https://www.mycore.de/classifications/author_roles#co_author"/>
</mods:role>
<mods:namePart type="family">Herbst</mods:namePart>
<mods:namePart type="given">Werner</mods:namePart>
</mods:name>
```

wird 
`<field name="mods.coAuthor">Herbst, Werner Herbst Werner</field>`

Bitte wenn vorhanden Display Name benutzen. Ansonsten aus Nachname und Vorname zusammenbauen.

Reply:
Kann ich mich darauf verlassen, dass entweder mods:displayForm oder mods:namePart-s in den mods-Daten zum mods:name vorhanden sind? Oder kann es sein, dass nur Text da ist?

Ausserdem habe ich gesehen, dass manchmal leere mods:affiliation-Tags im Solr-Dokument generiert werden. Dies würde ich dann bei dieser Gelegenheit auch korrigieren.

Reply:
displayForm sollte immer Vorhanden sein. Personen haben immer namePart Tags. 

Das die leeren affiliation Felder verschwinden ist gut aber nicht notwendig.